src/core/test/fuzzing/src/test/scala/Fuzzing.scala

```diff
@@ -26,10 +29,53 @@ case class TestObject[S <: PipelineStage](stage: S,
 
 }
 
-trait FuzzingMethods {
-  def compareDFs(df1: DataFrame, df2: DataFrame): Boolean = {
-    df1.collect().toSet == df2.collect().toSet
+trait FuzzingMethods extends TestBase {
+  val epsilon = 1e-4
+  implicit lazy val doubleEq: Equality[Double] = TolerantNumerics.tolerantDoubleEquality(epsilon)
+  implicit lazy val dvEq: Equality[DenseVector] = new Equality[DenseVector]{
+    def areEqual(a: DenseVector, b: Any): Boolean = b match {
+      case bArr:DenseVector =>
+        a.values.zip(bArr.values).forall {case (x, y) => doubleEq.areEqual(x, y)}
+    }
+  }
+
+  implicit lazy val rowEq: Equality[Row] = new Equality[Row]{
+    def areEqual(a: Row, bAny: Any): Boolean = bAny match {
+      case b:Row =>
+        if(a.length != b.length){return false}
```

Comment:
spacing is a little weird here (spaces before and after braces, parens)

Reply:
Done